core/src/main/scala/org/apache/spark/util/collection/PartitionedSerializedPairBuffer.scala

```diff
@@ -173,13 +181,13 @@ private[spark] class OrderedInputStream(metaBuffer: IntBuffer, kvBuffer: Chained
     if (metaBufferPos >= metaBuffer.position) {
       return -1
     }
-    val bytesRemainingInRecord = metaBuffer.get(metaBufferPos + VAL_END) - kvBufferPos
+    val bytesRemainingInRecord = metaBuffer.get(metaBufferPos + KEY_VAL_LEN)
     val toRead = math.min(bytesRemainingInRecord, len)
     kvBuffer.read(kvBufferPos, bytes, offs, toRead)
     if (toRead == bytesRemainingInRecord) {
       metaBufferPos += RECORD_SIZE
       if (metaBufferPos < metaBuffer.position) {
-        kvBufferPos = metaBuffer.get(metaBufferPos + KEY_START)
+        kvBufferPos = getKeyStartPos(metaBuffer, metaBufferPos)
       }
     } else {
       kvBufferPos += toRead
```

Comment:
Assume `bytesRemainingInRecord` is always 4. Image the following case:
```
read(bytes1, 0, 1);
read(bytes2, 0, 4);
```
The second `read` will read 3 bytes from the previous record, and 1 byte from another record. But the 1 byte may not belong to the next record.